### DISCUSSION TITLE
How can we make installing multiple versions of Umbraco friendlier

### DISCUSSION THREAD

**Discussion by JasonElkin** [umbraco/Umbraco-CMS#13127] [Features and ideas]
There is currently only one `Umbraco.Templates` dotnet template, and it's targeted to a specific version of Umbraco.

Right now I have `Umbraco.Templates::10.2.1` installed, which means that I can _only_ install Umbraco 10.2.1.

If I want to install a different version of Umbraco, I need to install a different version of the Template. This is a pain.

The whole point of the template is that the resulting project's code changes based on the options we specify. Is there any reason that this couldn't include the version of Umbraco itself in a friendly dropdown (instead of the text box, which doesn't really work)?

**Comment by jemayn** [comment]
It has a version flag though? If I run:
```
dotnet new umbraco -v 10.0.0 -n TestProject
```

I get a 10.0.0 site:
![image](https://user-images.githubusercontent.com/79840720/194074563-e56c8e66-6bb1-47da-ae26-99a893258dad.png)

**Reply by JasonElkin** [reply]
Yes, you get v10.0.0 added as a package reference. BUT the rest of the project files in the template are still for 10.2.1.

Right now for v10 that's just a few appsettings that I don't think will break anything, but if you choose version 9 💣

**Reply by JasonElkin** [reply]
So presumably any future release that requires changes to the project files could render the template unable to scaffold working projects using an older version of Umbraco.

Then there's the challenge of replicating bugs etc. if different templates versions scaffold different configurations etc.

**Reply by ronaldbarendse** [reply]
@JasonElkin Although it might not be explicitly documented, you should indeed install Umbraco using the template package of at least the same major version.

The [default .NET templates](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-new-sdk-templates) are also automatically updated with each SDK and some might have options that require a minimum SDK or framework version... This page also mentions the requirement of installing an older template if that's the case:

> Alternatively, to create a project that targets a framework earlier than the SDK that you're using, you might be able to do it by installing the NuGet package for the template. Common, web, and SPA project types use different packages per target framework moniker (TFM). For example, to create a console project that targets netcoreapp1.0, run dotnet new --install on Microsoft.DotNet.Common.ProjectTemplates.1.x.

So this isn't really an issue specific to Umbraco. Yes, we can make the templates completely backwards compatible, but that would complicate them even more and make maintaining it a real burden.

**Reply by JasonElkin** [reply]
Thanks @ronaldbarendse, that makes some sense.

Perhaps the Umbraco version property should be removed from the template?

I guess the difference between this and Microsoft's dotnet templates is that they are _just_ templates, whereas we're talking about _the_ official way to Install Umbraco. It feels like it should be easier to make the right choices here, and that installing any given version of Umbraco should really be a one-line command.

I've had a bit more of a look through the .NET docs and it's a bit disappointing, especially considering how other frameworks handle this so smoothly (e.g. `npm create package@version`).

**Reply by abjerner** [reply]
I think this is one of those moments where @prjseal deserves a 👍 

With his install script, you can get the commands to installing a specific version of Umbraco:

https://psw.codeshare.co.uk/?InstallUmbracoTemplate=true&UmbracoTemplateVersion=9.5.4&CreateSolutionFile=true&SolutionName=MySolution&UseUnattendedInstall=false&IncludeStarterKit=false

results in:

```
# Ensure we have the latest Umbraco templates
dotnet new -i Umbraco.Templates::9.5.4

# Create solution/project
dotnet new sln --name "MySolution"
dotnet new umbraco -n "MyProject"
dotnet sln add "MyProject"

dotnet run --project "MyProject"
#Running
```

https://psw.codeshare.co.uk/?InstallUmbracoTemplate=true&UmbracoTemplateVersion=10.2.1&CreateSolutionFile=true&SolutionName=MySolution&UseUnattendedInstall=false&IncludeStarterKit=false

results in:

```
# Ensure we have the latest Umbraco templates
dotnet new -i Umbraco.Templates::10.2.1

# Create solution/project
dotnet new sln --name "MySolution"
dotnet new umbraco -n "MyProject"
dotnet sln add "MyProject"

dotnet run --project "MyProject"
#Running
```

I use this for lack of something better. So as part of the commands is `dotnet new -i Umbraco.Templates::9.5.4`, it will replace the current template with the template for Umbraco 9.5.4, and later running `dotnet new -i Umbraco.Templates::10.2.1` will again replace the current template template for Umbraco 10.2.1.

The alternative would be if Umbraco it self had a CLI tool - eg. so you would be able to run something like `umbraco new --version 9.5.4`. But I imagine there is quite some work in building something like that.

**Reply by JasonElkin** [reply]
Thanks @abjerner, I looked at PSW but didn't realize you could specify the version - awesome!

I do like the idea of a CLI tool, but ideally would also have a solution that works with the Visual Studio template GUI too (maybe I'm asking too much 😬).

**Reply by prjseal** [reply]
> I think this is one of those moments where @prjseal deserves a 👍 
> 
> 
> 
> With his install script, you can get the commands to installing a specific version of Umbraco

Thanks mate. I'm glad you like it. I find it really useful for testing out the RC versions.

**Reply by deMD** [reply]
> Thanks @abjerner, I looked at PSW but didn't realize you could specify the version - awesome!
> 
> I do like the idea of a CLI tool, but ideally would also have a solution that works with the Visual Studio template GUI too (maybe I'm asking too much 😬).

This ☝️ 

But intil then, I give @prjseal tool a spin 😁

**Comment by sgartner-cityofboise** [comment]
> > I think this is one of those moments where @prjseal deserves a 👍
> > With his install script, you can get the commands to installing a specific version of Umbraco
> 
> Thanks mate. I'm glad you like it. I find it really useful for testing out the RC versions.

Thanks for the cool tool @prjseal!

This may not be your focus, but I was really hoping that it would let me switch major versions without having to go figure out what the most recent major version was.  For example, I was hoping this would work:
https://psw.codeshare.co.uk/?InstallUmbracoTemplate=true&TemplateVersion=10&CreateSolutionFile=true&SolutionName=MySolution&UseUnattendedInstall=false&IncludeStarterKit=false

Currently, the most recent version of v10 is 10.8.4, so I was hoping it would build the script for that version.  What it does instead is it ignores the provided version altogether and produces a script without a version indicator at all.

**Reply by prjseal** [reply]
I've added a new thing in so you can specify latest LTS version (Long Term Support)

So this script kind of does what you want now with the TemplateVersion set to LTS, although you can't specify v10 or v13 it will get you the latest LTS version which is what most people should use to be fair.

https://psw.codeshare.co.uk/?InstallUmbracoTemplate=true&TemplateVersion=LTS&CreateSolutionFile=true&SolutionName=MySolution&UseUnattendedInstall=false&IncludeStarterKit=false&TemplateName=Umbraco.Templates

**Comment by enkelmedia** [comment]
Personally I always use PSW to install some
version for testing etc. It’s so awesome! 

It might be nice if there could be a simple version of that on the our-page to make it even easier for people to install latest LTS or STS version. I guess that the “avarage” developer coming to Umbraco just wants to try the latest STS or LTS version not fiddle with specific versions.

**Reply by nul800sebastiaan** [reply]
There is, on each release 😅 Open for ideas to improve that one!

![image](https://github.com/umbraco/Umbraco-CMS/assets/304656/a2532592-5edf-434b-b3cb-62544a04ec97)

**Reply by deMD** [reply]
I ended up with a small command for my own devcli to run those command sequentially from a oneliner. I can now install all supported Umbraco version with the command `devcli Umbraco -v 13.0.0 -n projectname -d projectdir`. Works pretty well, only need to extend it to support all template options.

**Reply by nul800sebastiaan** [reply]
You can also write `dotnet new -i Umbraco.Templates::13.3.2 && dotnet new umbraco -n "MyProject && dotne run --project "MyProject"` - so join the commands with `&&` [https://psw.codeshare.co.uk/](https://psw.codeshare.co.uk/) includes an option for that as well. 

The point of keeping it multiple commands is to make it easier to know where the error is coming from if one of these commands fail.

**Reply by deMD** [reply]
Yeah that is basically what the command does under the hood, and it forwards all output so I can still see where something goes wrong 🙂 It's just that we regularly need to update Umbraco instances so having a short oneliner without duplicate names makes it a bit easier, for me a at least. Also the template creates the csproj, not the sln, so those commands also need to be added so with && it becomes quite the long command.

**Reply by nul800sebastiaan** [reply]
It does get long, as you can see in PSW as well, but that's okay, I don't have to type it manually, PSW just generates the output for me.

**Comment by enkelmedia** [comment]
@nul800sebastiaan Yeh, I know =D

If you "follow the path" of someone who has never seen Umbraco and try to get started.

https://our.umbraco.com/download/

The "Show install guidance" button takes you to the docs: https://docs.umbraco.com/umbraco-cms/v/13.latest-lts/fundamentals/setup/install

That says do run:

```
dotnet new install Umbraco.Templates
```

This is for v13 and since v14 is out - it would install v14.

I mean something simple like, when clicking the button "Show install guidance", a box with the snippet that you posted a screenshot of would be shown immediately for the latest version on that major.

![image](https://github.com/umbraco/Umbraco-CMS/assets/1782524/4cb7fe27-fca5-4a0a-829b-ec7fdfba45b5)

=D

**Reply by nul800sebastiaan** [reply]
You found a bug in the Docs! Not sure how dynamic we can make them though 🤔

**Reply by jemayn** [reply]
Since the docs are seperate files for each major, the command could just be changed to be fx

```
dotnet new install Umbraco.Templates::13.*
```

And so on for each major in the docs 🤔

**Reply by nul800sebastiaan** [reply]
Oooh yes, that's a great one @jemayn! I'll create an issue for them!

**Reply by nul800sebastiaan** [reply]
Docs have been fixed now! 🙌
